# **Tugas Praktikum**

Jalankan code berikut pada Google Colab dan PyDroid3 (Android Python) Application di Smartphone Android. Bandingkan hasilnya dan tuliskan analisa anda, tuliskan juga spesifikasi smartphone yang anda gunakan :).




In [ ]:
import numpy as np
import time
from sklearn.neighbors import NearestNeighbors
from annoy import AnnoyIndex
import hnswlib
import faiss
from sklearn.preprocessing import StandardScaler

# -------------------------------
# Contoh dataset kecil untuk testing
# -------------------------------
np.random.seed(42)
n_samples = 10000   # jumlah database vector
d = 128             # dimensi
X = np.random.random((n_samples, d)).astype('float32')

# Standarisasi fitur
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

k = 10  # jumlah nearest neighbors

# -------------------------------
# Exact NN (brute-force)
# -------------------------------
start = time.time()
nn = NearestNeighbors(n_neighbors=k, algorithm='brute', metric='euclidean')
nn.fit(X_scaled)
dist_exact, idx_exact = nn.kneighbors(X_scaled)
time_exact = time.time() - start
print(f"Exact NN done in {time_exact:.3f} s")

# -------------------------------
# Annoy
# -------------------------------
start = time.time()
f = X_scaled.shape[1]
index_annoy = AnnoyIndex(f, 'euclidean')
for i, v in enumerate(X_scaled):
    index_annoy.add_item(i, v)
index_annoy.build(10)
idx_annoy = [index_annoy.get_nns_by_vector(v, k) for v in X_scaled]
time_annoy = time.time() - start
print(f"Annoy done in {time_annoy:.3f} s")

# -------------------------------
# HNSW
# -------------------------------
start = time.time()
p_hnsw = hnswlib.Index(space='l2', dim=d)
p_hnsw.init_index(max_elements=n_samples, ef_construction=200, M=16)
p_hnsw.add_items(X_scaled)
p_hnsw.set_ef(200)
idx_hnsw, _ = p_hnsw.knn_query(X_scaled, k=k)
time_hnsw = time.time() - start
print(f"HNSW done in {time_hnsw:.3f} s")

# -------------------------------
# FAISS IVF
# -------------------------------
start = time.time()
quantizer = faiss.IndexFlatL2(d)
index_faiss = faiss.IndexIVFFlat(quantizer, d, 100, faiss.METRIC_L2)
index_faiss.train(X_scaled)
index_faiss.add(X_scaled)
index_faiss.nprobe = 10
_, idx_faiss = index_faiss.search(X_scaled, k)
time_faiss = time.time() - start
print(f"FAISS IVF done in {time_faiss:.3f} s")

# -------------------------------
# Tampilkan ringkasan waktu
# -------------------------------
print("\n=== Ringkasan Waktu (detik) ===")
print(f"Exact NN : {time_exact:.3f}")
print(f"Annoy    : {time_annoy:.3f}")
print(f"HNSW     : {time_hnsw:.3f}")
print(f"FAISS    : {time_faiss:.3f}")


Exact NN done in 2.004 s
Annoy done in 1.238 s
HNSW done in 9.534 s
FAISS IVF done in 0.467 s

=== Ringkasan Waktu (detik) ===
Exact NN : 2.004
Annoy    : 1.238
HNSW     : 9.534
FAISS    : 0.467


Spesifikasi Smartphone:
- Merk: POCO X3 PRO
- RAM: 8 GB
- PENYIMPANAN: 256 GB
- CHIPSET: Qualcomm Snapdragon 860

Hasil Run dari HP:

In [ ]:
# Exact NN done in 0.614 s
# Annoy done in 1.053 s
# HNSW done in 4.495 s

# === Ringkasan Waktu (detik) ===
# Exact NN : 0.614
# Annoy    : 1.053
# HNSW     : 4.495

#[Program finished]

Alasan saya tidak menggunakan FAISS adalah, karena saya telah mencoba untuk mendownload pipnya, baik secara quick install atau di terminal, tetapi hasilnya tetap nihil, kemungkinan besar karena tidak kompatibel atau library terlalu besar untuk digunakan pada pydroid

Berdasarkan analisis saya,

**PC**

Exact NN done in 2.004 s

Annoy done in 1.238 s

HNSW done in 9.534 s

FAISS IVF done in 0.467 s

=== Ringkasan Waktu (detik) ===

Exact NN : 2.004

Annoy    : 1.238

HNSW     : 9.534

FAISS    : 0.467

**Android**

Exact NN done in 0.614 s

Annoy done in 1.053 s

HNSW done in 4.495 s

=== Ringkasan Waktu (detik) ===

Exact NN : 0.614

Annoy    : 1.053

HNSW     : 4.495

Pada PC, yang paling cepat adalah FAISS, sementara itu di android yang paling cepat adalah Exact NN, yang kemungkinan dikarenakan dataset kecil, lalu terlihat juga kalau hasil kecepatan annoy di pc dan android tidak beda jauh sehingga bisa disimpulkan kalau annoy memiliki hasil yang paling stabil, lalu yang terakhir adalah HNSW, pada PC hnsw memiliki waktu yang paling lama, sementara itu di android hanya setengah waktu dari yang di pc, sehingga disimpulkan kalau hnsw di pc lebih lambat daripada di android, selain itu, kedua hnsw juga memiliki waktu yang paling lambat dibandingkan algoritma lain, tetapi memilik hasil yang paling memuaskan juga.